File: utils/paths.py
"""Canonical paths; construction never creates directories."""

from dataclasses import dataclass
import os
from pathlib import Path


def project_root() -> Path:
    return Path(
        os.environ.get("HAKUBOT_ROOT", Path(__file__).resolve().parents[1])
    ).resolve()


@dataclass(frozen=True)
class PluginPaths:
    plugin_id: str

    def __post_init__(self):
        if not self.plugin_id.isidentifier():
            raise ValueError("Invalid plugin ID")

    @property
    def config(self):
        return project_root() / "config/plugins" / self.plugin_id

    @property
    def data(self):
        if self.plugin_id == "identify":
            return project_root() / "plugins/identify/data"
        directory = {
            "groupmate_waifu": "waifu",
            "alive_stat": "alive_stats",
            "sk_predict": "sekai_cache",
        }.get(self.plugin_id, self.plugin_id)
        return project_root() / "data" / directory

    @property
    def cache(self):
        return project_root() / "cache" / self.plugin_id

    @property
    def resources(self):
        return project_root() / "plugins" / self.plugin_id / "resources"


def shared_cache(name: str) -> Path:
    if not name.isidentifier():
        raise ValueError("Invalid cache namespace")
    return project_root() / "cache/shared" / name

File: utils/paths.go
package utils

import (
	"fmt"
	"os"
	"path/filepath"
	"regexp"
)

var identifier = regexp.MustCompile(`^[A-Za-z_][A-Za-z0-9_]*$`)

type Paths struct{ Root string }

// NewPaths resolves the root once; constructing paths never creates directories.
func NewPaths(root string) (Paths, error) {
	if root == "" {
		root = os.Getenv("HAKUBOT_ROOT")
	}
	if root == "" {
		root = "."
	}
	absolute, err := filepath.Abs(root)
	return Paths{Root: absolute}, err
}

type PluginPaths struct{ Config, Data, Cache, Resources string }

func (p Paths) Plugin(id string) (PluginPaths, error) {
	if !identifier.MatchString(id) {
		return PluginPaths{}, fmt.Errorf("invalid plugin ID %q", id)
	}
	data := id
	switch id {
	case "groupmate_waifu":
		data = "waifu"
	case "alive_stat":
		data = "alive_stats"
	case "sk_predict":
		data = "sekai_cache"
	}
	dataPath := filepath.Join(p.Root, "data", data)
	if id == "identify" {
		dataPath = filepath.Join(p.Root, "plugins", id, "data")
	}
	return PluginPaths{filepath.Join(p.Root, "config", "plugins", id), dataPath, filepath.Join(p.Root, "cache", id), filepath.Join(p.Root, "plugins", id, "resources")}, nil
}
func (p Paths) SharedCache(name string) (string, error) {
	if !identifier.MatchString(name) {
		return "", fmt.Errorf("invalid cache namespace %q", name)
	}
	return filepath.Join(p.Root, "cache", "shared", name), nil
}
